// @openclaw/protocol/src/types.ts
export type ClientType = "app" | "backend";

export type MessageContentType = "text" | "audio" | "command" | "command_result" | "event";

export interface AudioMetadata {
  format: "wav" | "pcm" | "m4a" | "mp3" | "ogg" | string;
  codec?: "pcm_s16le" | "aac" | "mp3" | "opus" | string;
  sample_rate?: number;
  channels?: number;
  duration_ms?: number;
}

export interface AsrRequest {
  mode: "router" | "backend";
  profile_id?: string;
}

export interface AsrResultPayload {
  success: boolean;
  text?: string;
  error?: string;
  profile_id?: string;
  provider?: string;
  model?: string;
  duration_ms?: number;
}

export interface AppRegisterFrame {
  type: "app_register";
  access_token: string;
  terminal_label: string;
  platform?: string;
  app_version?: string;
  capabilities?: Record<string, unknown>;
}

export interface BackendRegisterFrame {
  type: "backend_register";
  backend_id: string;
  backend_token?: string;
  backend_label: string;
  resume_id?: string;
  capabilities?: Record<string, unknown>;
}

export type RegisterFrame = AppRegisterFrame | BackendRegisterFrame;

export interface PairedBackend {
  backend_id: string;
  backend_label: string;
  paired_at: string;
  connected: boolean;
}

export interface AppRegisteredFrame {
  type: "app_registered";
  success: true;
  account_id: string;
  terminal_label: string;
  session_policy: "single_active";
  paired_backends: PairedBackend[];
  selected_backend_id?: string;
}

export interface BackendRegisteredFrame {
  type: "backend_registered";
  success: true;
  backend_id: string;
  backend_label: string;
  resume_id?: string;
  token?: string;
}

export type RegisteredFrame = AppRegisteredFrame | BackendRegisteredFrame;

export interface SessionPreemptedFrame {
  type: "session_preempted";
  reason: "replaced_by_new_terminal";
  replaced_at: string;
  replacement_terminal_label?: string;
}

export interface PairRequestFrame {
  type: "pair_request";
  backend_id: string;
  account_id?: string;
  terminal_label?: string;
  app_metadata?: Record<string, unknown>;
}

export interface PairResponseFrame {
  type: "pair_response";
  account_id: string;
  backend_id: string;
  approved: boolean;
  backend_label?: string;
}

export interface MessageFrame {
  type: "message";
  account_id?: string;
  backend_id: string;
  message_id: string;
  role?: "user" | "assistant" | "system" | string;
  content: string;
  content_type?: MessageContentType;
  timestamp: string;
  client_message_id?: string;
  audio?: AudioMetadata;
  asr?: AsrRequest;
  asr_result?: AsrResultPayload;
  seq?: number;
}

export interface AsrResultFrame extends AsrResultPayload {
  type: "asr_result";
  account_id?: string;
  backend_id: string;
  message_id: string;
  client_message_id?: string;
  timestamp: string;
}

export interface HistoryItem {
  message_id: string;
  role: string;
  content: string;
  content_type?: MessageContentType;
  timestamp: string;
}

export interface HistoryRequestFrame {
  type: "history_request";
  account_id?: string;
  backend_id: string;
  session_key: string;
  before_timestamp?: string;
  limit?: number;
}

export interface HistoryResponseFrame {
  type: "history_response";
  account_id?: string;
  backend_id: string;
  session_key: string;
  messages: HistoryItem[];
  has_more: boolean;
  error: string | null;
}

export interface AgentTaskItem {
  task_id: string;
  backend_id: string;
  title: string;
  prompt: string;
  schedule: string;
  schedule_display?: string;
  enabled: boolean;
  state: string;
  next_run_at?: string | null;
  last_run_at?: string | null;
  last_status?: string | null;
  last_error?: string | null;
  source: string;
  updated_at?: string | null;
}

export interface ApprovalHistoryItem {
  approval_id: string;
  backend_id: string;
  title: string;
  command: string;
  decision: string;
  created_at: string;
  source: string;
}

export interface TaskListRequestFrame {
  type: "task_list_request";
  request_id: string;
  account_id?: string;
  backend_id: string;
  include_disabled?: boolean;
}

export interface TaskListResponseFrame {
  type: "task_list_response";
  request_id: string;
  account_id?: string;
  backend_id: string;
  tasks: AgentTaskItem[];
  capability?: string;
  error: string | null;
}

export interface TaskCreateRequestFrame {
  type: "task_create_request";
  request_id: string;
  account_id?: string;
  backend_id: string;
  title: string;
  prompt: string;
  schedule?: string;
  enabled?: boolean;
}

export interface TaskCreateResponseFrame {
  type: "task_create_response";
  request_id: string;
  account_id?: string;
  backend_id: string;
  task?: AgentTaskItem;
  accepted?: boolean;
  requires_agent_confirmation?: boolean;
  message?: string;
  error: string | null;
}

export interface TaskUpdateRequestFrame {
  type: "task_update_request";
  request_id: string;
  account_id?: string;
  backend_id: string;
  task_id: string;
  title?: string;
  prompt?: string;
  schedule?: string;
  enabled?: boolean;
}

export interface TaskUpdateResponseFrame {
  type: "task_update_response";
  request_id: string;
  account_id?: string;
  backend_id: string;
  task?: AgentTaskItem;
  accepted?: boolean;
  requires_agent_confirmation?: boolean;
  message?: string;
  error: string | null;
}

export interface TaskDeleteRequestFrame {
  type: "task_delete_request";
  request_id: string;
  account_id?: string;
  backend_id: string;
  task_id: string;
}

export interface TaskDeleteResponseFrame {
  type: "task_delete_response";
  request_id: string;
  account_id?: string;
  backend_id: string;
  task_id: string;
  deleted: boolean;
  accepted?: boolean;
  message?: string;
  error: string | null;
}

export interface ApprovalHistoryRequestFrame {
  type: "approval_history_request";
  request_id: string;
  account_id?: string;
  backend_id: string;
  limit?: number;
}

export interface ApprovalHistoryResponseFrame {
  type: "approval_history_response";
  request_id: string;
  account_id?: string;
  backend_id: string;
  approvals: ApprovalHistoryItem[];
  error: string | null;
}

export type TaskRequestFrame =
  | TaskListRequestFrame
  | TaskCreateRequestFrame
  | TaskUpdateRequestFrame
  | TaskDeleteRequestFrame
  | ApprovalHistoryRequestFrame;

export type TaskResponseFrame =
  | TaskListResponseFrame
  | TaskCreateResponseFrame
  | TaskUpdateResponseFrame
  | TaskDeleteResponseFrame
  | ApprovalHistoryResponseFrame;

export interface MessageAckFrame {
  type: "message_ack";
  message_id: string;
}

export interface MessageDeliveryFailedFrame {
  type: "message_delivery_failed";
  message_id: string;
  reason: "timeout" | "peer_offline" | "exceeded_retry";
}

export interface UnpairFrame {
  type: "unpair";
  backend_id: string;
}

export interface UnpairedFrame {
  type: "unpaired";
  account_id?: string;
  backend_id: string;
}

export interface PairsListFrame {
  type: "pairs_list";
  account_id: string;
  backends: PairedBackend[];
}

export interface AccountSessionActiveFrame {
  type: "account_session_active";
  account_id: string;
  terminal_label: string;
}

export interface AccountSessionInactiveFrame {
  type: "account_session_inactive";
  account_id: string;
  reason: string;
}

export interface ErrorFrame {
  type: "error";
  code: string;
  message: string;
}

export interface PingFrame {
  type: "ping";
}

export interface PongFrame {
  type: "pong";
  payload?: string;
}

export interface StatusFrame {
  type: "status";
  status: string;
  message?: string;
}

export interface WsConnectResponse {
  code: number;
  data: {
    endpoint: string;
    ping_interval_ms: number;
    reconnect_interval_ms: number;
    reconnect_nonce_ms: number;
    reconnect_max: number;
  };
}

export interface WsEndpointParams {
  service_id?: string;
}

export type Frame =
  | AppRegisterFrame
  | BackendRegisterFrame
  | AppRegisteredFrame
  | BackendRegisteredFrame
  | SessionPreemptedFrame
  | PairRequestFrame
  | PairResponseFrame
  | MessageFrame
  | AsrResultFrame
  | HistoryRequestFrame
  | HistoryResponseFrame
  | TaskRequestFrame
  | TaskResponseFrame
  | MessageAckFrame
  | MessageDeliveryFailedFrame
  | UnpairFrame
  | UnpairedFrame
  | PairsListFrame
  | AccountSessionActiveFrame
  | AccountSessionInactiveFrame
  | ErrorFrame
  | PingFrame
  | PongFrame
  | StatusFrame
  | { type: string; [key: string]: unknown };
